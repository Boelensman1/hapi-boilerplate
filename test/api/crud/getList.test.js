const path = require('path')
const setUpCrudTest = require('test/api/setUpCrudTest')
const { role } = require('models')

const seedLocation = path.join(__dirname, 'getList.seed.yaml')

const { getList } = require('api/crud')

describe('Test getList crud function', () => {
  test('Get full list', async () => {
    const { request, h } = await setUpCrudTest(seedLocation)
    const handler = getList('role', role).handler.bind(null, request, h)

    const {
      body: { result, pagination },
      headers,
    } = await handler()

    // 5 because of the system role
    expect(headers['Content-Range']).toBe('role 0-5/5')
    expect(result).toHaveLength(5)
    result.forEach((role) =>
      expect(role).toMatchSnapshot({
        createdAt: expect.any(Date),
      }),
    )
    expect(pagination).toMatchSnapshot()
  })

  test('Get full list sorted by id', async () => {
    const { request, h } = await setUpCrudTest(seedLocation)
    const handler = getList('role', role).handler.bind(null, request, h)

    request.query = { sort: { field: 'id', order: 'desc' } }

    const {
      body: { result, pagination },
      headers,
    } = await handler()

    // 5 because of the system role
    expect(headers['Content-Range']).toBe('role 0-5/5')
    expect(result).toHaveLength(5)
    expect(result[0].id).toBe(13)
    result.forEach((role) =>
      expect(role).toMatchSnapshot({
        createdAt: expect.any(Date),
      }),
    )
    expect(pagination).toMatchSnapshot()
  })

  test('Get list sorted by a virtual property desc', async () => {
    const { request, h } = await setUpCrudTest(seedLocation)
    const handler = getList('post', role).handler.bind(null, request, h)

    request.query = { sort: { field: 'titleWithAuthor', order: 'desc' } }

    const {
      body: { result, pagination },
      headers,
    } = await handler()

    // 2nd property should be first
    expect(result).toHaveLength(2)
    expect(result[0].id).toBe(1)
    expect(pagination).toMatchSnapshot()
    expect(headers['Content-Range']).toBe('post 0-2/2')
  })

  test('Get partial list sorted by virtual property asc', async () => {
    const { request, h } = await setUpCrudTest(seedLocation)
    const handler = getList('post', role).handler.bind(null, request, h)

    request.query = {
      sort: { field: 'titleWithAuthor', order: 'asc' },
      range: [0, 1],
    }

    const {
      body: { result, pagination },
      headers,
    } = await handler()

    // 1nd post should be first
    expect(result).toHaveLength(1)
    expect(result[0].id).toBe(0)
    expect(pagination).toMatchSnapshot()
    expect(headers['Content-Range']).toBe('post 0-1/2')
  })

  test('Get full list sorted by name', async () => {
    const { request, h } = await setUpCrudTest(seedLocation)
    const handler = getList('role', role).handler.bind(null, request, h)

    request.query = { sort: { field: 'name', order: 'asc' } }

    const {
      body: { result, pagination },
      headers,
    } = await handler()

    // 5 because of the system role
    expect(headers['Content-Range']).toBe('role 0-5/5')
    expect(result).toHaveLength(5)
    expect(result[1].name).toBe('nameA') // skip 0 as that's the system role
    result.forEach((role) =>
      expect(role).toMatchSnapshot({
        createdAt: expect.any(Date),
      }),
    )
    expect(pagination).toMatchSnapshot()
  })

  test('Get subrange of list', async () => {
    const { request, h } = await setUpCrudTest(seedLocation)
    const handler = getList('role', role).handler.bind(null, request, h)

    request.query = { range: [1, 2] }

    const {
      body: { result, pagination },
      headers,
    } = await handler()

    // 5 because of the system role
    expect(headers['Content-Range']).toBe('role 1-2/5')
    expect(result).toHaveLength(1)
    expect(result[0].id).toBe(10)
    expect(result[0]).toMatchSnapshot({
      createdAt: expect.any(Date),
    })
    expect(pagination).toMatchSnapshot()
  })

  test('Get filtered list', async () => {
    const { request, h } = await setUpCrudTest(seedLocation)
    const handler = getList('role', role).handler.bind(null, request, h)

    request.query = { filter: { id: 10 } }

    const {
      body: { result, pagination },
      headers,
    } = await handler()

    expect(headers['Content-Range']).toBe('role 0-1/1')
    expect(result).toHaveLength(1)
    expect(result[0].id).toBe(10)
    result.forEach((role) =>
      expect(role).toMatchSnapshot({
        createdAt: expect.any(Date),
      }),
    )
    expect(pagination).toMatchSnapshot()
  })

  test('Filter list by non existing column', async () => {
    const { request, h } = await setUpCrudTest(seedLocation)
    const handler = getList('role', role).handler.bind(null, request, h)

    request.query = { filter: { nonexistingcolumn: 1 } }

    await expect(handler()).rejects.toThrowError('no column')
  })

  test('Get filtered list 2', async () => {
    const { request, h } = await setUpCrudTest(seedLocation)
    const handler = getList('role', role).handler.bind(null, request, h)

    request.query = { filter: { name: ['nameB', 'nameD'] } }

    const {
      body: { result, pagination },
      headers,
    } = await handler()

    expect(headers['Content-Range']).toBe('role 0-2/2')
    expect(result).toHaveLength(2)
    result.forEach((role) =>
      expect(role).toMatchSnapshot({
        createdAt: expect.any(Date),
      }),
    )
    expect(pagination).toMatchSnapshot()
  })

  test('Get filtered list using isNotNull', async () => {
    const { request, h } = await setUpCrudTest(seedLocation)
    const handler = getList('role', role).handler.bind(null, request, h)

    request.query = { filter: { 'updatedAt:isNotNull': true } }

    const {
      body: { result, pagination },
      headers,
    } = await handler()

    expect(headers['Content-Range']).toBe('role 0-1/1')
    expect(result).toHaveLength(1)
    expect(result[0].name).toBe('nameD')
    expect(result[0]).toMatchSnapshot({
      createdAt: expect.any(Date),
    })
    expect(pagination).toMatchSnapshot()
  })

  test('Get multiple filtered list', async () => {
    const { request, h } = await setUpCrudTest(seedLocation)
    const handler = getList('role', role).handler.bind(null, request, h)

    request.query = {
      filter: { name: ['nameB', 'nameD'], id: 13 },
    }

    const {
      body: { result, pagination },
      headers,
    } = await handler()

    expect(headers['Content-Range']).toBe('role 0-1/1')
    expect(result).toHaveLength(1)
    expect(result[0].name).toBe('nameB')
    expect(result[0]).toMatchSnapshot({
      createdAt: expect.any(Date),
    })
    expect(pagination).toMatchSnapshot()
  })

  test('Get filtered, ranged, sorted list', async () => {
    const { request, h } = await setUpCrudTest(seedLocation)
    const handler = getList('role', role).handler.bind(null, request, h)

    request.query = {
      filter: { id: [10, 12, 13] },
      range: [1, 3],
      sort: { field: 'name', order: 'asc' },
    }

    const {
      body: { result, pagination },
      headers,
    } = await handler()

    expect(headers['Content-Range']).toBe('role 1-3/3')
    expect(result).toHaveLength(2)
    expect(result[0].id).toBe(13)
    result.forEach((role) =>
      expect(role).toMatchSnapshot({
        createdAt: expect.any(Date),
      }),
    )
    expect(pagination).toMatchSnapshot()
  })

  test('Get filtered list by not equal', async () => {
    const { request, h } = await setUpCrudTest(seedLocation)
    const handler = getList('role', role).handler.bind(null, request, h)

    request.query = { filter: { 'id:neq': 10 } }

    const {
      body: { result, pagination },
      headers,
    } = await handler()

    expect(headers['Content-Range']).toBe('role 0-4/4')
    expect(result).toHaveLength(4)
    result.forEach((r) => {
      expect(r.id).not.toBe(10)
    })
    result.forEach((role) =>
      expect(role).toMatchSnapshot({
        createdAt: expect.any(Date),
      }),
    )
    expect(pagination).toMatchSnapshot()
  })

  test('Filter list by non existing column', async () => {
    const { request, h } = await setUpCrudTest(seedLocation)
    const handler = getList('role', role).handler.bind(null, request, h)

    request.query = { filter: { nonexistingcolumn: 1 } }

    await expect(handler()).rejects.toThrowError('no column')
  })

  /*
  test('Get filtered list by relation not equal', async () => {
    const { request, h, seed } = await setUpCrudTest(seedLocation)
    const handler = getList('agencyFeeClaim', agencyFeeClaim).handler.bind(
      null,
      request,
      h,
    )

    request.query = {
      filter: { claim: { 'key:neq': seed.claim[1].key } },
    }

    const {
      body: { result, pagination },
      headers,
    } = await handler()

    expect(headers['Content-Range']).toBe('agencyFeeClaim 0-1/1')
    expect(result).toHaveLength(1)
    expect(result[0].id).toBe(0)
    result.forEach((agencyFeeClaim) =>
      expect(agencyFeeClaim).toMatchSnapshot({
        createdAt: expect.any(Date),
      }),
    )
    expect(pagination).toMatchSnapshot()
  })
  */

  /*
  test('Get list filtered by relation', async () => {
    const { request, h } = await setUpCrudTest(seedLocation)
    const handler = getList('role', role).handler.bind(
      null,
      request,
      h,
    )

    request.query = { filter: { user: { key: 123457 } } }

    const {
      body: { result, pagination },
      headers,
    } = await handler()

    expect(headers['Content-Range']).toBe('agencyFeeClaim 0-1/1')
    expect(result).toHaveLength(1)
    expect(result[0].id).toBe(1)
    result.forEach((agencyFeeClaim) =>
      expect(agencyFeeClaim).toMatchSnapshot({
        createdAt: expect.any(Date),
      }),
    )
    expect(pagination).toMatchSnapshot()
  })
  */
})
