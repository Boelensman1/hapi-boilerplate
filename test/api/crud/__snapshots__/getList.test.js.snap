// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test getList crud function Get filtered list 1`] = `
Object {
  "createdAt": Any<Date>,
  "id": 10,
  "name": "nameC",
  "posts_create": 0,
  "posts_delete": 0,
  "posts_read": 0,
  "posts_update": 0,
  "roles_create": 0,
  "roles_delete": 0,
  "roles_read": 0,
  "roles_update": 0,
  "updatedAt": null,
  "users_create": 0,
  "users_delete": 0,
  "users_read": 0,
  "users_update": 0,
}
`;

exports[`Test getList crud function Get filtered list 2`] = `
Object {
  "count": 1,
}
`;

exports[`Test getList crud function Get filtered list 2 1`] = `
Object {
  "createdAt": Any<Date>,
  "id": 13,
  "name": "nameB",
  "posts_create": 0,
  "posts_delete": 0,
  "posts_read": 0,
  "posts_update": 0,
  "roles_create": 0,
  "roles_delete": 0,
  "roles_read": 0,
  "roles_update": 0,
  "updatedAt": null,
  "users_create": 0,
  "users_delete": 0,
  "users_read": 0,
  "users_update": 0,
}
`;

exports[`Test getList crud function Get filtered list 2 2`] = `
Object {
  "createdAt": Any<Date>,
  "id": 11,
  "name": "nameD",
  "posts_create": 0,
  "posts_delete": 0,
  "posts_read": 0,
  "posts_update": 0,
  "roles_create": 0,
  "roles_delete": 0,
  "roles_read": 0,
  "roles_update": 0,
  "updatedAt": 1970-01-01T00:00:01.000Z,
  "users_create": 0,
  "users_delete": 0,
  "users_read": 0,
  "users_update": 0,
}
`;

exports[`Test getList crud function Get filtered list 2 3`] = `
Object {
  "count": 2,
}
`;

exports[`Test getList crud function Get filtered list using $hasProp 1`] = `
Object {
  "createdAt": Any<Date>,
  "id": 11,
  "name": "nameD",
  "posts_create": 0,
  "posts_delete": 0,
  "posts_read": 0,
  "posts_update": 0,
  "roles_create": 0,
  "roles_delete": 0,
  "roles_read": 0,
  "roles_update": 0,
  "updatedAt": 1970-01-01T00:00:01.000Z,
  "users_create": 0,
  "users_delete": 0,
  "users_read": 0,
  "users_update": 0,
}
`;

exports[`Test getList crud function Get filtered list using $hasProp 2`] = `
Object {
  "count": 1,
}
`;

exports[`Test getList crud function Get filtered, ranged, sorted list 1`] = `
Object {
  "createdAt": Any<Date>,
  "id": 13,
  "name": "nameB",
  "posts_create": 0,
  "posts_delete": 0,
  "posts_read": 0,
  "posts_update": 0,
  "roles_create": 0,
  "roles_delete": 0,
  "roles_read": 0,
  "roles_update": 0,
  "updatedAt": null,
  "users_create": 0,
  "users_delete": 0,
  "users_read": 0,
  "users_update": 0,
}
`;

exports[`Test getList crud function Get filtered, ranged, sorted list 2`] = `
Object {
  "createdAt": Any<Date>,
  "id": 10,
  "name": "nameC",
  "posts_create": 0,
  "posts_delete": 0,
  "posts_read": 0,
  "posts_update": 0,
  "roles_create": 0,
  "roles_delete": 0,
  "roles_read": 0,
  "roles_update": 0,
  "updatedAt": null,
  "users_create": 0,
  "users_delete": 0,
  "users_read": 0,
  "users_update": 0,
}
`;

exports[`Test getList crud function Get filtered, ranged, sorted list 3`] = `
Object {
  "count": 3,
  "previous": "https://www.example.org/crud/role/?filter=%7B%22id%22%3A%5B10%2C12%2C13%5D%7D&range=%5B0%2C1%5D&sort=%7B%22field%22%3A%22name%22%2C%22order%22%3A%22asc%22%7D",
}
`;

exports[`Test getList crud function Get full list 1`] = `
Object {
  "createdAt": Any<Date>,
  "id": 1,
  "name": "SYSTEM",
  "posts_create": 0,
  "posts_delete": 0,
  "posts_read": 0,
  "posts_update": 0,
  "roles_create": 0,
  "roles_delete": 0,
  "roles_read": 0,
  "roles_update": 0,
  "updatedAt": null,
  "users_create": 0,
  "users_delete": 0,
  "users_read": 0,
  "users_update": 0,
}
`;

exports[`Test getList crud function Get full list 2`] = `
Object {
  "createdAt": Any<Date>,
  "id": 10,
  "name": "nameC",
  "posts_create": 0,
  "posts_delete": 0,
  "posts_read": 0,
  "posts_update": 0,
  "roles_create": 0,
  "roles_delete": 0,
  "roles_read": 0,
  "roles_update": 0,
  "updatedAt": null,
  "users_create": 0,
  "users_delete": 0,
  "users_read": 0,
  "users_update": 0,
}
`;

exports[`Test getList crud function Get full list 3`] = `
Object {
  "createdAt": Any<Date>,
  "id": 11,
  "name": "nameD",
  "posts_create": 0,
  "posts_delete": 0,
  "posts_read": 0,
  "posts_update": 0,
  "roles_create": 0,
  "roles_delete": 0,
  "roles_read": 0,
  "roles_update": 0,
  "updatedAt": 1970-01-01T00:00:01.000Z,
  "users_create": 0,
  "users_delete": 0,
  "users_read": 0,
  "users_update": 0,
}
`;

exports[`Test getList crud function Get full list 4`] = `
Object {
  "createdAt": Any<Date>,
  "id": 12,
  "name": "nameA",
  "posts_create": 0,
  "posts_delete": 0,
  "posts_read": 0,
  "posts_update": 0,
  "roles_create": 0,
  "roles_delete": 0,
  "roles_read": 0,
  "roles_update": 0,
  "updatedAt": null,
  "users_create": 0,
  "users_delete": 0,
  "users_read": 0,
  "users_update": 0,
}
`;

exports[`Test getList crud function Get full list 5`] = `
Object {
  "createdAt": Any<Date>,
  "id": 13,
  "name": "nameB",
  "posts_create": 0,
  "posts_delete": 0,
  "posts_read": 0,
  "posts_update": 0,
  "roles_create": 0,
  "roles_delete": 0,
  "roles_read": 0,
  "roles_update": 0,
  "updatedAt": null,
  "users_create": 0,
  "users_delete": 0,
  "users_read": 0,
  "users_update": 0,
}
`;

exports[`Test getList crud function Get full list 6`] = `
Object {
  "count": 5,
}
`;

exports[`Test getList crud function Get full list sorted by id 1`] = `
Object {
  "createdAt": Any<Date>,
  "id": 13,
  "name": "nameB",
  "posts_create": 0,
  "posts_delete": 0,
  "posts_read": 0,
  "posts_update": 0,
  "roles_create": 0,
  "roles_delete": 0,
  "roles_read": 0,
  "roles_update": 0,
  "updatedAt": null,
  "users_create": 0,
  "users_delete": 0,
  "users_read": 0,
  "users_update": 0,
}
`;

exports[`Test getList crud function Get full list sorted by id 2`] = `
Object {
  "createdAt": Any<Date>,
  "id": 12,
  "name": "nameA",
  "posts_create": 0,
  "posts_delete": 0,
  "posts_read": 0,
  "posts_update": 0,
  "roles_create": 0,
  "roles_delete": 0,
  "roles_read": 0,
  "roles_update": 0,
  "updatedAt": null,
  "users_create": 0,
  "users_delete": 0,
  "users_read": 0,
  "users_update": 0,
}
`;

exports[`Test getList crud function Get full list sorted by id 3`] = `
Object {
  "createdAt": Any<Date>,
  "id": 11,
  "name": "nameD",
  "posts_create": 0,
  "posts_delete": 0,
  "posts_read": 0,
  "posts_update": 0,
  "roles_create": 0,
  "roles_delete": 0,
  "roles_read": 0,
  "roles_update": 0,
  "updatedAt": 1970-01-01T00:00:01.000Z,
  "users_create": 0,
  "users_delete": 0,
  "users_read": 0,
  "users_update": 0,
}
`;

exports[`Test getList crud function Get full list sorted by id 4`] = `
Object {
  "createdAt": Any<Date>,
  "id": 10,
  "name": "nameC",
  "posts_create": 0,
  "posts_delete": 0,
  "posts_read": 0,
  "posts_update": 0,
  "roles_create": 0,
  "roles_delete": 0,
  "roles_read": 0,
  "roles_update": 0,
  "updatedAt": null,
  "users_create": 0,
  "users_delete": 0,
  "users_read": 0,
  "users_update": 0,
}
`;

exports[`Test getList crud function Get full list sorted by id 5`] = `
Object {
  "createdAt": Any<Date>,
  "id": 1,
  "name": "SYSTEM",
  "posts_create": 0,
  "posts_delete": 0,
  "posts_read": 0,
  "posts_update": 0,
  "roles_create": 0,
  "roles_delete": 0,
  "roles_read": 0,
  "roles_update": 0,
  "updatedAt": null,
  "users_create": 0,
  "users_delete": 0,
  "users_read": 0,
  "users_update": 0,
}
`;

exports[`Test getList crud function Get full list sorted by id 6`] = `
Object {
  "count": 5,
}
`;

exports[`Test getList crud function Get full list sorted by name 1`] = `
Object {
  "createdAt": Any<Date>,
  "id": 1,
  "name": "SYSTEM",
  "posts_create": 0,
  "posts_delete": 0,
  "posts_read": 0,
  "posts_update": 0,
  "roles_create": 0,
  "roles_delete": 0,
  "roles_read": 0,
  "roles_update": 0,
  "updatedAt": null,
  "users_create": 0,
  "users_delete": 0,
  "users_read": 0,
  "users_update": 0,
}
`;

exports[`Test getList crud function Get full list sorted by name 2`] = `
Object {
  "createdAt": Any<Date>,
  "id": 12,
  "name": "nameA",
  "posts_create": 0,
  "posts_delete": 0,
  "posts_read": 0,
  "posts_update": 0,
  "roles_create": 0,
  "roles_delete": 0,
  "roles_read": 0,
  "roles_update": 0,
  "updatedAt": null,
  "users_create": 0,
  "users_delete": 0,
  "users_read": 0,
  "users_update": 0,
}
`;

exports[`Test getList crud function Get full list sorted by name 3`] = `
Object {
  "createdAt": Any<Date>,
  "id": 13,
  "name": "nameB",
  "posts_create": 0,
  "posts_delete": 0,
  "posts_read": 0,
  "posts_update": 0,
  "roles_create": 0,
  "roles_delete": 0,
  "roles_read": 0,
  "roles_update": 0,
  "updatedAt": null,
  "users_create": 0,
  "users_delete": 0,
  "users_read": 0,
  "users_update": 0,
}
`;

exports[`Test getList crud function Get full list sorted by name 4`] = `
Object {
  "createdAt": Any<Date>,
  "id": 10,
  "name": "nameC",
  "posts_create": 0,
  "posts_delete": 0,
  "posts_read": 0,
  "posts_update": 0,
  "roles_create": 0,
  "roles_delete": 0,
  "roles_read": 0,
  "roles_update": 0,
  "updatedAt": null,
  "users_create": 0,
  "users_delete": 0,
  "users_read": 0,
  "users_update": 0,
}
`;

exports[`Test getList crud function Get full list sorted by name 5`] = `
Object {
  "createdAt": Any<Date>,
  "id": 11,
  "name": "nameD",
  "posts_create": 0,
  "posts_delete": 0,
  "posts_read": 0,
  "posts_update": 0,
  "roles_create": 0,
  "roles_delete": 0,
  "roles_read": 0,
  "roles_update": 0,
  "updatedAt": 1970-01-01T00:00:01.000Z,
  "users_create": 0,
  "users_delete": 0,
  "users_read": 0,
  "users_update": 0,
}
`;

exports[`Test getList crud function Get full list sorted by name 6`] = `
Object {
  "count": 5,
}
`;

exports[`Test getList crud function Get list sorted by a virtual property desc 1`] = `
Object {
  "count": 2,
}
`;

exports[`Test getList crud function Get multiple filtered list 1`] = `
Object {
  "createdAt": Any<Date>,
  "id": 13,
  "name": "nameB",
  "posts_create": 0,
  "posts_delete": 0,
  "posts_read": 0,
  "posts_update": 0,
  "roles_create": 0,
  "roles_delete": 0,
  "roles_read": 0,
  "roles_update": 0,
  "updatedAt": null,
  "users_create": 0,
  "users_delete": 0,
  "users_read": 0,
  "users_update": 0,
}
`;

exports[`Test getList crud function Get multiple filtered list 2`] = `
Object {
  "count": 1,
}
`;

exports[`Test getList crud function Get partial list sorted by virtual property asc 1`] = `
Object {
  "count": 2,
  "next": "https://www.example.org/crud/post/?sort=%7B%22field%22%3A%22titleWithAuthor%22%2C%22order%22%3A%22asc%22%7D&range=%5B1%2C2%5D",
}
`;

exports[`Test getList crud function Get subrange of list 1`] = `
Object {
  "createdAt": Any<Date>,
  "id": 10,
  "name": "nameC",
  "posts_create": 0,
  "posts_delete": 0,
  "posts_read": 0,
  "posts_update": 0,
  "roles_create": 0,
  "roles_delete": 0,
  "roles_read": 0,
  "roles_update": 0,
  "updatedAt": null,
  "users_create": 0,
  "users_delete": 0,
  "users_read": 0,
  "users_update": 0,
}
`;

exports[`Test getList crud function Get subrange of list 2`] = `
Object {
  "count": 5,
  "next": "https://www.example.org/crud/role/?range=%5B2%2C3%5D",
  "previous": "https://www.example.org/crud/role/?range=%5B0%2C1%5D",
}
`;
